### PR TITLE
Feature/6 enc emb

### LINKED ISSUE #6: [PRE] [Pipeline] Embeddings

### LINKED ISSUE #47: [PRE] [Pipeline] Encodings
This covers the core encoding stage before embeddings:
vocabulary building
token-to-id conversion
sequence padding/truncation
label-to-id conversion

### PR DESCRIPTION
**Pull request by aashutosh-chapagain**
Closes #48 
Closes #6 

This PR introduces modular encoding and embedding components along with comprehensive unit tests. It establishes a clean and extensible pipeline from tokenized text -> encoded sequences -> embedding representations, supporting both baseline and transformer-based approaches.

1. **Encoding Module**
Implemented:
- `VocabEncoder`
- `SequenceEncoder`
- `LabelEncoder`
Supports:
Vocabulary building with frequency filtering
Token -> ID conversion with <pad> and <unk>
Sequence padding/truncation and attention mask generation
Label encoding/decoding

2. **Embedding Module (Modular Backends)**

2.1 **_Scratch Embedding Backend_**
- `TrainableEmbeddingBackend` using nn.Embedding
- Mask-aware mean pooling for sentence representation
- Configurable via `ScratchEmbeddingConfig`

2.2 **_BERT Embedding Backend_**
- `BertTokenizerWrapper` for preprocessing
- `BertEmbeddingBackend` using Hugging Face AutoModel
- Supports:
 - CLS / mean pooling
 - fine-tuning toggle
- Configurable via `BertEmbeddingConfig`

2.3 **_Shared Interface_**
- Introduced `EmbeddingOutput` for unified outputs:
 - `token_embeddings`
 - `attention_mask`
 - `sentence_embedding`
- Enables interchangeable embedding backends for downstream models

3. **Unit Tests**
- Added structured pytest-based tests:
 - `test_encoding.py` -> encoding logic and edge cases
 - `test_embedding_scratch.py` -> embedding correctness, masking, pooling
 
[todo] `test_bert.py` -> build mocked transformer 
[todo] Optionally extend with `FastText` or other embedding strategies

### CODE REVIEW

**Comment by aashutosh-chapagain** [comment]
vocab encoder usage
vocab-encoder

sequence encoder usage
sequence-encoder

label encoder usage
label-encoder

baseline embedding usage
baseline-embedding-scratch-usage

bert usage - use raw- conservatively cleaned descriptions
bert usage

**Comment by AndreHrs** [comment]
I don't think it's a good idea to put example code as in image. Probably try to use the code block here in the comment.
But good job.
~Ps: there's a code conflict so you can try to fix that and it should be ready to go!~
I resolved the code conflict, looks like it's just a CI dependency issue. Nothing much. 
Just need to wait for the test to pass.

**Review by AndreHrs — APPROVED** [review]
Code looks okay, unit tests are pretty thorough too

**Comment by aashutosh-chapagain** [comment]
The test has successfully passed now. Also the code blocks was too much hassle to quote here, and thus the images. Below is the markdown friendly script from my notebook.

```

processor = OneTextPreProcessor(
    keep_numbers=False,
    column_map=column_map,
    lemmatize=False
)

train_proc = processor.pre_process_df(train_df, text_col)
valid_proc = processor.pre_process_df(valid_df, text_col)
test_proc = processor.pre_process_df(test_df, text_col)

token_col = f"{text_col}_tokens"

train_tokens = train_proc[token_col].tolist()
valid_tokens = valid_proc[token_col].tolist()
test_tokens = test_proc[token_col].tolist()

```

```

vocab_encoder = VocabEncoder(min_freq=2)
vocab_encoder.fit(train_tokens)
```

```


sequence_encoder = SequenceEncoder(
    vocab_encoder=vocab_encoder,
    max_length=max_length,
    padding="post",
    truncating="post",
)

X_train = sequence_encoder.encode_sequences(train_tokens)
X_valid = sequence_encoder.encode_sequences(valid_tokens)
X_test = sequence_encoder.encode_sequences(test_tokens)

train_masks = sequence_encoder.build_attention_masks(X_train)
valid_masks = sequence_encoder.build_attention_masks(X_valid)
test_masks = sequence_encoder.build_attention_masks(X_test)
```

```


label_encoder = LabelEncoder()
label_encoder.fit(train_labels)

y_train = label_encoder.encode_many(train_labels)
y_valid = label_encoder.encode_many(valid_labels)
y_test = label_encoder.encode_many(test_labels)
```

```


import torch
from modules.embedding import ScratchEmbeddingConfig, TrainableEmbeddingBackend

scratch_config = ScratchEmbeddingConfig(
    vocab_size=vocab_encoder.vocab_size,
    embedding_dim=100,
    pad_idx=vocab_encoder.pad_id,
    dropout=0.1,
)

scratch_backend = TrainableEmbeddingBackend(scratch_config)

X_train_tensor = torch.tensor(X_train, dtype=torch.long)
scratch_output = scratch_backend(X_train_tensor[:4])
```

```


from modules.embedding import (
    BertEmbeddingConfig,
    BertTokenizerWrapper,
    BertEmbeddingBackend,
)

bert_config = BertEmbeddingConfig(
    model_name="bert-base-uncased",
    max_length=160,
    dropout=0.1,
    pooling="mean",
    fine_tune=True,
)

bert_tokenizer = BertTokenizerWrapper(bert_config)
bert_backend = BertEmbeddingBackend(bert_config)

texts = train_proc[text_col].astype(str).tolist()[:4]
bert_inputs = bert_tokenizer.encode_texts(texts)

bert_output = bert_backend(
    input_ids=bert_inputs["input_ids"],
    attention_mask=bert_inputs["attention_mask"],
    token_type_ids=bert_inputs.get("token_type_ids"),
)
```
